File: Leetcode/cpp/medium/56.cpp
/*
 * @lc app=leetcode id=56 lang=cpp
 *
 * [56] Merge Intervals
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;
        ranges::sort(intervals,
                     [](const auto& a, const auto& b) { return a[0] < b[0]; });
        for (const auto& interval : intervals) {
            int l = interval[0], r = interval[1];
            if (ans.empty() || l > ans.back()[1])
                ans.push_back({l, r});
            else
                ans.back()[1] = max(ans.back()[1], r);
        }
        return ans;
    }
};
// @lc code=end

File: Leetcode/rust/medium/56.rs
/*
 * @lc app=leetcode id=56 lang=rust
 *
 * [56] Merge Intervals
 */

// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
impl Solution {
    pub fn merge(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        intervals.sort_by_key(|interval| interval[0]);
        let mut ans: Vec<Vec<i32>> = Vec::new();
        for interval in intervals {
            let l = interval[0];
            let r = interval[1];
            if ans.is_empty() || l > ans.last().unwrap()[1] {
                ans.push(interval);
            } else {
                ans.last_mut().unwrap()[1] = r.max(ans.last().unwrap()[1]);
            }
        }
        ans
    }
}
// @lc code=end
